### ISSUE TITLE
Support a typed API?

### ISSUE THREAD

**Issue by cauthmann** [spacejam/sled#1266]
Hello,

#### Use Case:

The current API isn't typed - as the docs say, it's a `HashMap<[u8],[u8]>`, not a `HashMap<K,V>`. I can easily write a `u32`, then try to read the same value as a JSON encoded custom type.

While type safety cannot be guaranteed at the database level (on disk, it's just a bunch of bytes), we could give the user an API closer to `HashMap<K,V>` for `Tree`s. Not only would this prevent (some) errors, it could also allow automatic conversion, reducing boilerplate.

#### Proposed Change:

Any such API must have two constraints:
  * The user must be able to specify not just a type, but also how to convert the type to bytes and back.
  * if keys and values are specified as `[u8] without conversions` (the default), the API must be backwards compatible

I've written a little proof of concept and tested it against several encoding schemes here:
https://github.com/cauthmann/sled-typed/blob/main/src/lib.rs
I've implemented this for `[u8]`, `str`, ints (converted to big endian, for keys), anything serializable with serde and types from the zerocopy crate. That should be representative for both zero copy types and types that require a serialization step.

My proof of concept satisfies the first constraint (at least for a subset of the Tree API), but not the second: an untyped tree only accepts `&[u8]`, not any `AsRef<[u8]>` type as before. Currently it won't accept `IVec<T>` in place of `T` either.

Before investigating further, I wanted to check if a typed API is something you might be interested in upstreaming? If so, I'll move the changes into sled itself and try to figure out if it can be made backwards compatible. Otherwise I'll keep this as a minimal out-of-tree facade for my own needs.

#### Who Benefits From The Change(s)?

Anyone who would like to use Trees as a generic `HashMap<K, V>` instead of just `HashMap<[u8], [u8]>`.

#### Alternative Approaches

Keep the database agnostic and expect the users to build their own APIs around it.

**Comment by spacejam** [comment]
Hey @cauthmann, this is an excellent proposal! Right off the bat, it addresses the major concerns that have caused me to delay implementing better typed tree or typed value support:

* avoids depending on or reexporting serde, zerocopy, or json-related crates
* allows for flexible ownership so that you can do things like convert integer types to big endian before storing them

So, the 1.0 version of sled is about to be published very soon (tomorrow night). It's OK to break things in this release, but breakage must have an eye toward the next 5 years and seem reasonable given that time horizon. I have a flurry of things to address over the next few hours, but I'll be thinking about this and looking for ways of incorporating it.

**Comment by cauthmann** [comment]
Just documenting all the types that would need to be touched.

  * `IVec<T>` to add `.decode() -> T` and maybe `From<T>` (though that trait may require specialization)

These just need to be typed to forward the types to a contained or returned type.
  * `Event<K, V>` to return `IVec<K>`/`IVec<V>`
  * `Subscriber<K, V>` to return `Event<K, V>`
  * `Iter<T>` to return `IVec<T>`
  * `CompareAndSwapError<V>` to contain `IVec<V>`

These need to be typed to accept IVecs of the correct format. Better yet, they would get a different API to accept the correct types directly, e.g. `insert(K, V)` instead of `insert(K: AsRef<[u8]>, V: IVec)`
  * `Tree<K, V>`
  * `TransactionalTree<K, V>`
  * `Batch<K, V>`.
 
Anytime we want to change a method's signature, we need a different type, e.g. `Tree` vs `TypedTree`. I have not found a way to override these otherwise, except to rename them to `.insert_typed()` or similar.

This can either be a completely separate type, but it should be fine to just wrap the existing types (possibly with a `Deref` implementation to the wrapped type), like I did in my proof of concept.

For all the others, we have to make a decision:
  * create two types (either separate, or wrappers)
  * add type parameters with a default (which is a [non-breaking change](https://github.com/rust-lang/rfcs/blob/master/text/1105-api-evolution.md#minor-change-adding-a-defaulted-type-parameter))

I believe that type parameters are the better choice as they cause less code duplication.

In other words, I think all of this can be done post-1.0 without breaking changes.

**Comment by MDeiml** [comment]
Some thoughts about supporting caching the deserialized values in the future:
* Either methods have to return something like `Arc`s or values have to be copied for every read
* Types would need to decide if they should be cached in value form (like serde types) or not (like `[u8]`/ zerocopy types)
* In some cases types would benefit from being cached in value form but still borrow some of the underlying bytes
For example look at the following struct:
```rust
struct Hybrid<'a> {
    a: ComplicatedOwnedInnerStruct,
    b: &'a [u8],
}
```
For something like this it would be very hard if not impossible to track ownership so I would suggest not allowing them.

**Comment by MDeiml** [comment]
Actually if you would rewrite this struct as
```rust
struct Hybrid {
    a: ComplicatedOwnedInnerStruct,
    b: IVec,
}
```
You would not have to track ownership, as it is tracked by the `Arc` inside the `IVec`.

**Comment by MDeiml** [comment]
I've written a prototype of the `Encoder` trait which would also support future caching. I can share it if you want. Also I would be happy to implement this into the API if it's ok with you, @spacejam.

**Comment by cauthmann** [comment]
Hello @MDeiml,

there are a few reasons why I didn't consider caching the decoded value:

It's not very useful in the general case:
  * decoding/deserializing smallish buffers is generally fast, compared to the `.get()` call that precedes it (in my benchmarks, <20% of the time was spent deserializing, even with expensive JSON).

It's difficult to implement without introducing additional costs:
  * Not every `IVec` gets decoded (for example the return value of `.insert()` or anything that returns a key). As such, we'd have to implement decoding on demand, which would require a mutex around the decode call.
  * Returning the cached values would require either wrapping them in an `Arc` or returning copies.
  * Such a cache would have to be opt-in per tree, or it'll more than double memory usage when using owned structures.

There is no unique deserialization for an IVec:
  * I may want to decode into a zero-copy read-only structure for performance, but I may also want to decode into a mutable owned structure, make some changes, then write it back into the database. Or I may want to skip parsing the JSON value entirely and just forward the string into a HTTP response. I believe that it would be useful to allow the user to override the decoder via turbofish: `ivec.decode::<MyOtherDecoder>()`.
  * Even if we were to enforce a single decoded type, we cannot stop the user from opening a tree twice, with different encoders.
  * There may be no way at all to decode an `IVec`. As the encode and decode logic don't share any code, I believe we should do two separate traits here, so we don't force the user to write decoding logic for keys.

All that being said, I'd be interested in seeing your approach. If we can prepare for such a cache without introducing costs or complications, then it may be a worthwhile change, and we can test the costs and benefits of an actual cache implementation later.


I have already implemented some of this inside sled, but decided to wait until the pre-1.0-churn has settled before continuing, to avoid merge conflicts. Though if you want to take this off my plate, that's fine with me too.

**Comment by MDeiml** [comment]
Those are some great points I didn't consider. I originally just assumed caching would be a plan because I remember someone mentioning it in some other issue (which I can't find now). In summary the pros and cons of caching deserialized values would be

Pros:
* (A bit) faster access when reading the same value multiple times
* Avoid serializing / deserializing values that get written and then read immediately

Cons:
* Almost impossible to implement without forcing the user to keep a reference to the `Tree` object and even then caches could probably not be shared between multiple handles of the same `Tree`. (You would have to have some kind of `HashMap<Type, Cache>` which I don't think can be implemented in rusts type system)
* Either values get decoded that are not read, or you have some kind of "lazy" decoding, which wouldn't play nice with multithreading.

So all in all I would say that caching deserialized values is not something that would work well in a multithreading context. Caching also could be added later without end users having to change any code. Only encoders would have to be updated to choose a caching strategy.

Your code in https://github.com/cauthmann/sled-typed/blob/main/src/lib.rs would work fine then. The only thing I would change is to let encoders decide if they want to return an `IVecWrapper` as the default encoder would want to return an `IVec` and the zerocopy encoder could save some time by only checking the layout once.

Do you have a fork where you implemented this into sled itself?

**Comment by MDeiml** [comment]
Actually sled behaves more like a `BTreeMap` than a `HashMap` meaning entries are ordered by their byte representation. So as long as you ensure that the byte representation has the same ordering your values probably would be iterated in sorted order.

**Comment by ckaran** [comment]
@MDeiml yes, but even that isn't enough.  See #1171 for more info, but the crux of the issue is that there are types which can serve as keys, but which don't have a good lexicographic sort order.  With big rationals, you always have a numerator and a denominator, usually stored in some reduced form.  We know (as humans) that the correct order for `1/3`, `1/2`, and `2/3` is that order.  However, a lexicographic sort order will yield `1/2`, `1/3`, `2/3`.  You can try to map to some normalized form, but that can be **very** time and space intensive (e.g., denormalize all keys so that they all have some common denominator, *and hope that you never insert/change a key that forces you to denormalize again*).  This is the reason why I was wondering if it is possible to implement something that keeps them in sorted order.

**Comment by cauthmann** [comment]
I meant to wait until 1.0, but then I got bored and went ahead anyway:

https://github.com/cauthmann/sled

Basic inserts/gets/deletes work. There's [a testcase](https://github.com/cauthmann/sled/blob/main/tests/test_typed_api.rs) showing off the new API.

Reviews and comments welcome. 

There is a distinction between the old `IVec`, which exposes bytes in a backward compatible manner, and `IVec<Encoding>`, which is meant to be accessed using the Encoder/Decoder.

So far, a two questions:

**When an IVec has an Encoding, do we keep exposing the bytes using implicit APIs?**
I did implement `AsRef` and `AsRefMut`. I deliberately did not implement `Deref`, and I did not implement byte-wise comparisons via `PartialEq` and friends.

I'm starting to think that `AsRef` was a mistake, and we should hide byte-access behind a new method `.as_bytes() -> [u8]`. If the user wants to circumvent the type-safe API, let them be explicit about it.

This change would be a bit more invasive, because `AsRef` and `AsMut` are used internally in a few places. Opinions?

b) Internally, everything stays an `IVec`, but the typed API has to return them as `IVec<Encoding>`. For that, I've added a method to change the type. This shouldn't clone the IVec, it shouldn't touch any refcounts, it should be a noop. But I'm not sure if [my magic is correct](https://github.com/cauthmann/sled/blob/b122be23c037525ba43dece1a7b648a6ca2e8d92/src/ivec.rs#L304-L307). If you're going to review one function, make it this one.

**Comment by ckaran** [comment]
@cauthmann, the API is mainly at https://github.com/cauthmann/sled/blob/main/src/encoding.rs, correct?  I would like to suggest that encoding be fallible like decoding is (there is an Error type, and encoding returns a Result).  There's a lot of weird corner cases that could turn up, along with possible future headaches, so it's better to expose it now.

Also, all errors should implement [`std::error::Error`](https://doc.rust-lang.org/std/error/trait.Error.html).  That will make it easier to create MUCH clearer error messages in the future.

> I'm starting to think that `AsRef` was a mistake, and we should hide byte-access behind a new method `.as_bytes() -> [u8]`. If the user wants to circumvent the type-safe API, let them be explicit about it.

Would that be an `unsafe` method?

Haven't reviewed the code yet, and am unlikely to, I barely have time to review things in my inbox as it is... :cry:

**Comment by cauthmann** [comment]
Thanks for your feedback, @ckaran.

> I would like to suggest that encoding be fallible like decoding is (there is an Error type, and encoding returns a Result). There's a lot of weird corner cases that could turn up, along with possible future headaches, so it's better to expose it now.

I'm not sure. Of all the serialization frameworks I've looked at, I encountered only two failure modes:
  * memory allocation problems, but those panic anyway
  * serde is used to serialize something not supported by the format, e.g. serializing an u128 in a format which does not support u128. That is not something that can be fixed at runtime.

There is a real cost of making encoding fallible. `tree.insert(key, value)` encodes `key` and `value` and would need to be able to return `sled::Error` and either of the encoder errors. I am honestly not sure how to do that without breaking backwards compatibility and/or over-complicating the API.

Can you point out any use case where encoding failure is expected?

As a workaround, those use cases can always use the old API, or do the fallible part as an intermediate manual step, e.g. `tree.insert(1, my_stuff.try_encode()?)?`.

> Also, all errors should implement [`std::error::Error`](https://doc.rust-lang.org/std/error/trait.Error.html). That will make it easier to create MUCH clearer error messages in the future.

Yes, good point, I will do that. I forgot about `std::convert::Infallible` to opt out of it.

> Would that be an `unsafe` method?

There's nothing `unsafe` (in the rust sense) about fiddling with the bytes of an IVec, it's just that the User probably doesn't intend to when using the new API.

**Comment by ckaran** [comment]
> Can you point out any use case where encoding failure is expected?

- Objects or object graphs that can be in either a canonical form or an intermediate form.  Converting to the canonical form (or verifying that it is already in that form) may be expensive, so it might only be done by request.  The encoder might not be able to (or might not want to) force the conversion, so it returns an error to warn the user that conversion is necessary before it can be serialized out, at which point the user can make the decision to for canonicalization or not (could be useful for time sensitive tasks).  
- Encoding of objects that are actually proxies for other objects.  The proxy might not be able to be serialized except under certain circumstances, like a packet capture proxy might be communicating with another device that is slow, etc.

> There's nothing `unsafe` (in the rust sense) about fiddling with the bytes of an IVec, it's just that the User probably doesn't intend to when using the new API.

OK, so this would just be a convention with a big warning banner on it, right?  Sounds good to me.

**Comment by cauthmann** [comment]
Implementing mutation or network communication in an `Encoder` seems like a stretch. When you call `.insert`, you should have insertable data, not some random object that may or may not be serializable at some point in the future.

For those cases, I believe it is cleaner to gather the data before the database access:
```rust
   let canonical_graph = graph.canonicalize()?;
   tree.insert(key, canonical_graph);
   let proxy_state = proxy.wait_for_data()?;
   tree.insert(key, proxy_state);
```
Both might be good candidates to employ the typestate pattern.

Made some progress today:
  * Batch insertion works
  * compare_and_swap and friends work
 
Next up:
  * Iterators and Ranges
  * Subscribers and Events
  * Transactions. Multi-tree transactions with mixed tree types are going to be fun.
  * Possibly redo everything based on the lessons I've learned along the way.

**Comment by tv42** [comment]
One more concern I haven't seen mentioned yet: merge operators.

The typed values for `Tree::merge` may be different from the type for values, *or* it may be the same.

Example of same: counter, `u64` and `u64` is plenty.

Example of different: base type with just a "change list" passed to `Tree::merge`.
